src/LinkAdder.php

```diff
@@ -0,0 +1,115 @@
+<?php
+
+namespace Spatie\Crawler;
+
+use Tree\Node\Node;
+use GuzzleHttp\Psr7\Uri;
+use InvalidArgumentException;
+use Psr\Http\Message\UriInterface;
+use Symfony\Component\DomCrawler\Link;
+use Symfony\Component\DomCrawler\Crawler as DomCrawler;
+
+class LinkAdder
+{
+    /** @var \Spatie\Crawler\Crawler */
+    protected $crawler;
+
+    public function __construct(Crawler $crawler)
+    {
+        $this->crawler = $crawler;
+    }
+
+    public function addFromHtml(string $html, UriInterface $foundOnUrl)
+    {
+        $allLinks = $this->extractLinksFromHtml($html, $foundOnUrl);
+
+        collect($allLinks)
+            ->filter(function (UriInterface $url) {
+                return $this->hasCrawlableScheme($url);
+            })
+            ->map(function (UriInterface $url) {
+                return $this->normalizeUrl($url);
+            })
+            ->each(function (UriInterface $url) use ($foundOnUrl) {
```

Comment:
These early returns could be refactored to a `filter` chain. Sounds cleaner to me, but might look a but messy in practice.

Reply:
Seems cleaner, indeed. Though the last check on maximumCrawlCountReached() must be in the each loop. I've refactored the rest though.

Reply:
To clarify intent there, you could do `->take($this->cawler->remainingCrawlCount())`